File: python/models/alumno.py
from .persona import Persona


class Alumno(Persona):

    def __init__(self, cedula, nombre_completo, correo, programa):
        super().__init__(cedula, nombre_completo, correo)
        self.notas = []
        self.programa = programa

    def agregar_nota(self, nota):
        if len(self.notas) < 3:
            self.notas.append(nota)
            return True

        print("El alumno ya posee el máximo de 3 notas.")
        return False

    def eliminar_ultima_nota(self):
        if self.notas:
            return self.notas.pop()

        print("No hay notas registradas.")
        return None

    def calcular_promedio(self):
        if not self.notas:
            return 0

        return sum(self.notas) / len(self.notas)

    def esta_aprobado(self):
        if len(self.notas) != 3:
            return False

        return self.programa.evaluar_aprobacion(self.notas)

File: java/Alumno.java
import java.util.ArrayList;
import java.util.List;

public class Alumno extends Persona {
    private final ProgramaAcademico programa;
    private final List<Double> notas;

    public Alumno(String cedula, String nombre, String correo, ProgramaAcademico programa) {
        this(cedula, nombre, correo, programa, new ArrayList<>());
    }

    public Alumno(String cedula, String nombre, String correo, ProgramaAcademico programa, List<Double> notas) {
        super(cedula, nombre, correo);
        this.programa = programa;
        this.notas = new ArrayList<>(notas);
    }

    public ProgramaAcademico getPrograma() {
        return programa;
    }

    public List<Double> getNotas() {
        return List.copyOf(notas);
    }

    public void agregarNota(double nota) {
        if (notas.size() >= 3) {
            throw new ValidationException("Cada alumno puede tener máximo 3 notas.");
        }
        notas.add(Validaciones.nota(nota));
    }

    public double eliminarUltimaNota() {
        if (notas.isEmpty()) {
            throw new ValidationException("El alumno no tiene notas para deshacer.");
        }
        return notas.remove(notas.size() - 1);
    }

    public double calcularPromedio() {
        return notas.stream().mapToDouble(Double::doubleValue).average().orElse(0);
    }

    public boolean estaAprobado() {
        return programa.evaluarAprobacion(notas);
    }

    public String obtenerEstatus() {
        if (notas.size() < 3) {
            return "PENDIENTE";
        }
        return estaAprobado() ? "APROBADO" : "REPROBADO";
    }

    public String toFileLine() {
        List<Double> valores = new ArrayList<>(notas);
        while (valores.size() < 3) {
            valores.add(0.0);
        }
        return String.join(",", getCedula(), getNombre(), getCorreo(), programa.getNombre(),
                Formato.numero(valores.get(0)), Formato.numero(valores.get(1)), Formato.numero(valores.get(2)));
    }
}
